kernel/src/events/observer.rs

```diff
@@ -30,10 +32,65 @@ use super::Events;
 /// assert!(empty.upgrade().is_empty());
 /// ```
 pub trait Observer<E: Events>: Send + Sync {
-    /// Notify the observer that some interesting events happen.
+    /// Notifies the observer that some interesting events happen.
     fn on_events(&self, events: &E);
 }
 
 impl<E: Events> Observer<E> for () {
-    fn on_events(&self, events: &E) {}
+    fn on_events(&self, _events: &E) {}
+}
+
+/// A set of observers.
+///
+/// This is similar to [`Subject`] and can also notify observers of events. However, there are two
+/// key differences:
+///  * This type does not have any inner locks. Therefore, observers are allowed to break the
+///    atomic mode as far as this type is concerned.
+///  * This type has no support for [`EventsFilter`].
+///
+/// [`Subject`]: super::Subject
+/// [`EventsFilter`]: super::EventsFilter
+pub struct ObserverSet<E: Events>(BTreeSet<KeyableWeak<dyn Observer<E>>>);
```

Comment:
Could this struct be renamed to `LockfreeSubject` to emphasize the distinction from `Subject`?

Reply:
Actually it needs to be used inside a lock, e.g., `Mutex<LockfreeSubject>`. Is it expected? Is `Mutex<LockedSubject>` more natural?

Reply:
Emm, I now realize that `LockfreeSubject` is confusing, since it doesn't provide any interior mutability. `LockedSubject` also feels like a Guard type, even though the type can be accessed directly via `&mut self`.

On second thought, the current `LockedSubject` might deserve the name `Subject` because it's the more basic form. The original `Subject`, which contains an internal lock, could be renamed `SyncSubject` (a synchronized subject) or `GuardedSubject` (a subject that provides interior mutability via an internal lock). What's your opinion about it?

Reply:
I cannot think of a better name than `SyncSubject`. The only issue is that `Subject` is also `Sync`. Having an immutable reference `&Subject` is fine, though it is useless because all of the methods require `&mut self`.